### ISSUE TITLE
DNNLinearCombinedClassifier weight_column_name doesn't work in v1.0

### ISSUE THREAD

**Issue by erko** [tensorflow/tensorflow#8043]
After migrating to tensorflow v1.0 of my sources, weight_column_name in estimators (tested in tf.contrib.learn.DNNLinearCombinedClassifier) seems to not working now.
Scripts now outputs this error:
`ERROR:tensorflow:Could not create metric ops for MetricSpec(metric_fn=_predictions_streaming_mean, prediction_key=logistic, label_key=None, weight_key=class_weights)`

### Environment info
Operating System: 
CentOS Linux release 7.3.1611 (Core)

Docker image within TF package:
1. Docker latest image: tensorflow/tensorflow
Created container:
`docker run -it -p 8888:8888 -v ..... --memory=12g tensorflow/tensorflow`
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`:
`1.0.0`

### Reproducing

1. Download wide_n_deep_tutorial.py: [https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/learn/wide_n_deep_tutorial.py](url)
(This solely works fine)
2. Then add weightings (insert lines at):
line 26 `import numpy as np`
line 130 `weight_column_name='class_weights',`
line 153 `feature_cols['class_weights'] = tf.constant(np.ones(np.shape(df[LABEL_COLUMN].values)))`
3. Code executes with error:
`ERROR:tensorflow:Could not create metric ops for MetricSpec(metric_fn=_predictions_streaming_mean, prediction_key=logistic, label_key=None, weight_key=class_weights)`
Full trace of error is below.

If we repeat this procedures with version 0.10, code works well.

I tried to reshape weight feature col to [2, n], to [n, 2] - also, doesn't help.

### Trace
```
ERROR:tensorflow:Could not create metric ops for MetricSpec(metric_fn=_predictions_streaming_mean, prediction_key=logistic, label_key=None, weight_key=class_weights).
Traceback (most recent call last):
  File "wide_n_deep_tutorial.py", line 236, in <module>
    tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/platform/app.py", line 44, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "wide_n_deep_tutorial.py", line 199, in main
    FLAGS.train_data, FLAGS.test_data)
  File "wide_n_deep_tutorial.py", line 188, in train_and_eval
    m.fit(input_fn=lambda: input_fn(df_train), steps=train_steps)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/deprecation.py", line 280, in new_func
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 426, in fit
    loss = self._train_model(input_fn=input_fn, hooks=hooks)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 934, in _train_model
    model_fn_ops = self._call_legacy_get_train_ops(features, labels)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1003, in _call_legacy_get_train_ops
    train_ops = self._get_train_ops(features, labels)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1162, in _get_train_ops
    return self._call_model_fn(features, labels, model_fn_lib.ModeKeys.TRAIN)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1133, in _call_model_fn
    model_fn_results = self._model_fn(features, labels, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/dnn_linear_combined.py", line 312, in _dnn_linear_combined_model_fn
    features, labels, mode, _make_training_op, logits=logits)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 531, in head_ops
    labels, predictions)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 1390, in _eval_metric_ops
    return estimator._make_metrics_ops(metrics, features, labels, predictions)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 258, in _make_metrics_ops
    result[name] = metric.create_metric_ops(features, labels, predictions)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/metric_spec.py", line 220, in create_metric_ops
    weights=inputs[self.weight_key])
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 1440, in _predictions_streaming_mean
    return metrics_lib.streaming_mean(predictions, weights=weights)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/contrib/metrics/python/ops/metric_ops.py", line 392, in streaming_mean
    updates_collections=updates_collections, name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/metrics_impl.py", line 262, in mean
    math_ops.to_float(weights), values)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/weights_broadcast_ops.py", line 166, in broadcast_weights
    with ops.control_dependencies((assert_broadcastable(weights, values),)):
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/weights_broadcast_ops.py", line 102, in assert_broadcastable
    weights_rank_static))
```
### Update 3/10/2017
Somehow in process of editing/formatting this post, forgot one line at the end of traceback:
`ValueError: weights can not be broadcast to values. values.rank=2. weights.rank=1.`

**Comment by tatatodd** [comment]
Thanks for filing the issue @erko 

@martinwicke might know about this, or might know who can investigate why `tf.contrib.learn.DNNLinearCombinedClassifier(weight_column_name)` isn't working correctly.

**Comment by honkentuber** [comment]
This is a bug. Workaround is to reshape weights to (batch, 1).

e.g.,
feature_cols["class_weights"] = tf.reshape(
      tf.ones_like(df[LABEL_COLUMN].values), shape=(-1, 1))

**Comment by erko** [comment]
@honkentuber Thanks for help,
but not works this way too:
without reshape, error message: `ValueError: weights can not be broadcast to values. values.rank=2. weights.rank=1.`
with reshape: `ValueError: weights can not be broadcast to values. values.rank=1. weights.rank=2.`
- tried exactly your line of code, and some variations, results are same. Looks like problem is where weights.rank is adapting to values.rank

**Comment by honkentuber** [comment]
Sorry, I wasn't running against the right version. Also, add this line just after `label` is created in `input_fn`:
label = tf.reshape(label, shape=(-1, 1))

**Comment by erko** [comment]
@honkentuber thanks, you absolutely right. Reshaping label too (w class_weight) helps to solve this problem.

**Comment by afhuertass** [comment]
Hi ! 

i'm having the same issue. but changing the shape of the label tensor didn't work . i define my metrics as:
```
METRICS = {
    'loss' : metric_spec.MetricSpec (
        metric_fn = metric_ops.streaming_mean , 
        prediction_key = 'loss' ,
        weight_key = None ,
        label_key = None 
    )
}
```
and pass it to the eval_metrics argument, creating a Experiment object. but i i run it i get the same error, 
**ValueError: weights can not be broadcast to values. values.rank=0. weights.rank=2**

i'm using a batch size of 2 ( for debuging ) . But i find it really weird because i specify **weight_key = None**, and by checking the code, it says that in that case the weights arre just 1.  Some how a "weights" parameter is reaching streaming_mean , even if i said that i dont want it to be weigthed. 
The output of my  **input_fn** is a list   [ images , labels   ]

So.. what i'm doing wrong? 
Thanks  in advance :)

**Comment by honkentuber** [comment]
I believe the problem here is that the streaming_mean function doesn't take
a label argument, so it's mistaking labels for weights. You can
replace metric_ops.streaming_mean
with:

def _predictions_streaming_mean(labels, predictions, weights):
return metric_ops.streaming_mean(values=predictions, weights=weights)


Philip Tucker |  Software Engineer |  ptucker@google.com |

**Comment by afhuertass** [comment]
Yeah 

i ended up solving it by doing something similar, "masking" the streaming_mean function, and passint to it weights  = tf.ones( shape= () , dtype=tf.float32 )

that did the trink ;) thanks.

**Comment by honkentuber** [comment]
FYI, this should be fixed in 1.1.


Philip Tucker |  Software Engineer |  ptucker@google.com |

**Comment by ntvy95** [comment]
Hello... @honkentuber 

Tensorflow version: 1.2.1
Anaconda version: 1.6.3
OS: Windows 8.1 Pro

I also run into the same problem with DNNClassifer: 

* Logging error:

> Message: 'Weights {} has shape {}, expanding to make it 2d.'
> Arguments: ('sample_weight', TensorShape([Dimension(60480)]))

* ValueError:

> ValueError: weights can not be broadcast to values. values.rank=1. weights.rank=2. values.shape=(60480,). weights.shape=(60480, 1).

I have fixed according to your instruction but later I get this error (which I believe this is due to the fact that 'sample_weight' was forced to the shape(60480,1)):

> ValueError: Features are incompatible with given information. Given features: {'is_signed_out': <tf.Tensor 'Const:0' shape=(60480,) dtype=bool>, 'timesend_start_mapping_hour': <tf.Tensor 'Const_3:0' shape=(60480,) dtype=int64>, 'urls_per_cookie': <tf.Tensor 'Const_8:0' shape=(60480,) dtype=int64>, 'timesend_start_mapping_minute': <tf.Tensor 'Const_4:0' shape=(60480,) dtype=int64>, 'geoip_lon': <tf.Tensor 'Const_2:0' shape=(60480,) dtype=float64>, 'res_width': <tf.Tensor 'Const_5:0' shape=(60480,) dtype=float64>, 'sample_weight': <tf.Tensor 'Reshape_1:0' shape=(60480, 1) dtype=float64>, 'ua_os_family': <tensorflow.python.framework.sparse_tensor.SparseTensor object at 0x000000000C7670B8>, 'geoip_lat': <tf.Tensor 'Const_1:0' shape=(60480,) dtype=float64>, 'res_height': <tf.Tensor 'Const_6:0' shape=(60480,) dtype=float64>, 'urlrefs_per_cookie': <tf.Tensor 'Const_7:0' shape=(60480,) dtype=float64>}, required signatures: {'is_signed_out': TensorSignature(dtype=tf.bool, shape=TensorShape([Dimension(60480)]), is_sparse=False), 'timesend_start_mapping_hour': TensorSignature(dtype=tf.int64, shape=TensorShape([Dimension(60480)]), is_sparse=False), 'geoip_lon': TensorSignature(dtype=tf.float64, shape=TensorShape([Dimension(60480)]), is_sparse=False), 'res_width': TensorSignature(dtype=tf.float64, shape=TensorShape([Dimension(60480)]), is_sparse=False), 'timesend_start_mapping_minute': TensorSignature(dtype=tf.int64, shape=TensorShape([Dimension(60480)]), is_sparse=False), 'ua_os_family': TensorSignature(dtype=tf.string, shape=None, is_sparse=True), 'geoip_lat': TensorSignature(dtype=tf.float64, shape=TensorShape([Dimension(60480)]), is_sparse=False), 'sample_weight': TensorSignature(dtype=tf.float64, shape=TensorShape([Dimension(60480)]), is_sparse=False), 'res_height': TensorSignature(dtype=tf.float64, shape=TensorShape([Dimension(60480)]), is_sparse=False), 'urls_per_cookie': TensorSignature(dtype=tf.int64, shape=TensorShape([Dimension(60480)]), is_sparse=False), 'urlrefs_per_cookie': TensorSignature(dtype=tf.float64, shape=TensorShape([Dimension(60480)]), is_sparse=False)}.

What should I do now?

Thank you very much in advance!

**Comment by brightbytes-dude** [comment]
This is still not fixed in TensorFlow 1.3.0.

I'm using the high-level API only, and define the weights column in exactly the same fashion as I define other feature columns:

`tf.contrib.layers.real_valued_column(self.schema.label_weight_column)`

... where I have tests that prove that `self.schema.label_weight_column` is the string naming my weight column `"outcome_class_weight"`

I then define the model as follows, where `deep_feature_columns` includes `tf.contrib.layers.real_valued_column(self.schema.label_weight_column)`:

    tf.contrib.learn.DNNClassifier(
      model_dir=self.schema.model_dir,
      feature_columns=deep_feature_columns,
      weight_column_name=self.schema.label_weight_column,
      hidden_units=self.schema.hidden_units,
      activation_fn=self.schema.activation_fn,
      dropout=self.schema.dropout_probability,
      gradient_clip_norm=self.schema.gradient_clip_norm_ratio,
      enable_centered_bias=self.schema.enable_centered_bias,
      optimizer=optimizer,
      config=run_config
    )

Finally, upon running, I see the weights broadcast Exception:

```
Traceback (most recent call last):
  File "runner.py", line 89, in <module>
    tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "runner.py", line 40, in main
    metrics = trainer.train_and_test()
  File "/Users/aaron/git/bb_tensorflow/domain/model_trainer.py", line 93, in train_and_test
    monitors=[self.__build_monitor__()]
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 296, in new_func
    return func(*args, **kwargs)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 458, in fit
    loss = self._train_model(input_fn=input_fn, hooks=hooks)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 958, in _train_model
    model_fn_ops = self._get_train_ops(features, labels)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1165, in _get_train_ops
    return self._call_model_fn(features, labels, model_fn_lib.ModeKeys.TRAIN)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1136, in _call_model_fn
    model_fn_results = self._model_fn(features, labels, **kwargs)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/estimators/dnn.py", line 209, in _dnn_model_fn
    logits=logits)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 854, in create_model_fn_ops
    enable_centered_bias=self._enable_centered_bias)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 651, in _create_model_fn_ops
    weighted_average_loss, predictions, labels, weight_tensor)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 908, in _metrics
    _indicator_labels_streaming_mean(labels, weights))
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 1937, in _indicator_labels_streaming_mean
    weights = weights_broadcast_ops.broadcast_weights(weights, labels)
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/python/ops/weights_broadcast_ops.py", line 167, in broadcast_weights
    with ops.control_dependencies((assert_broadcastable(weights, values),)):
  File "/Users/aaron/git/bb_tensorflow/lib/python3.6/site-packages/tensorflow/python/ops/weights_broadcast_ops.py", line 103, in assert_broadcastable
    weights_rank_static, values.shape, weights.shape))
ValueError: weights can not be broadcast to values. values.rank=1. weights.rank=2. values.shape=(18018,). weights.shape=(18018, 1).
```

I'm going to attempt @honkentuber 's proposed fix, but I strongly suspect I'll encounter @ntvy95 's exception after doing so.  (Will update if not.)

Also, apologies for being forward, but why was this ticket closed when the exception still occurs in the most recent stable version of TensorFlow?

**Comment by martinwicke** [comment]
Is this a problem with `tf.estimator.Estimator`? I hope not, and I believe that that is why this issue was closed.

FYI @ispirmustafa

**Comment by ispirmustafa** [comment]
contrib DNNClassifier doesn't support a weight column. it expects that weight is processed within the input_fn properly.
core DNNClassifier (tf.estimator.DNNClassifier) accepts a weight column. Please use that instead.

**Comment by brightbytes-dude** [comment]
I was able to get this working by riffing off the suggested fix, and changing only the label definition provided by the input function for all my models as per this git diff:

```
-    label = tf.constant(
-      df[self.schema.label_column].values,
-      dtype=tf.bool
+    label = tf.reshape(
+      tf.constant(
+        df[self.schema.label_column].values,
+        dtype=tf.bool
+      ),
+      [-1, 1]
     )
```

So, while I would hope that the Estimator would take care of this for me, it's not clear whether or not its responsibility is to take care of it for me. :-) 

IOW, I'm lucky I was able to suss this relatively quickly based upon @honkentuber 's comments: others might not be so fortunate, so some fix would be cool.  Just not sure where that would be.

Finally, while I have you ;-) , two quick questions that you can send me to Stack Overflow for if you don't want to deal:

I stupidly created my dev/test sets with the weight column.  But, obviously, for tuning and testing, I don't want the weight column in my features or defined by the model.  (I tried it, and the presence of the weight column garnered an accuracy of 99.4% because the weight column is completely correlated with the label). And, clearly, when making predictions, defining that column is right out.

So, the first question is: Will it be possible to load a trained model (via its `model_dir`) into an Estimator definition that doesn't define the weight column in the features or as the `weight_column_name` param?  (If not, then what exactly is the utility of the `weight_column_name` param?)

I'm going to try that soon, but can't immediately due to the the following question:

I'm using the (deprecated) Early Stopping ValidationMonitor feature described in "Logging and Monitoring" tutorial ... that appears to have been taken down recently. (?!)  The problem is that the Early Stopping ValidationMonitor is passed into the model during training, and thus the input function it uses for test data may (?) need to provide the same features for the dev/test set as it does for the training set.  

The question then is: does its input function need to match the signature of the training input function?  

If so, I'll (tragically!) need to stop using this feature, which I love because it narrows the wide degree of freedom of manually tuning `training_steps`.  IOW, tuning `early_stopping_steps` can occur over a much narrower range of values than tuning `training_steps`, which I dig.

**Comment by brightbytes-dude** [comment]
@ispirmustafa - Ok, I'll try that, but according top https://www.tensorflow.org/api_docs/python/tf/contrib/learn/DNNClassifier#__init__ the `weight_column_name` field is supported.

**Comment by ispirmustafa** [comment]
I agree `weight_column_name` is misleading. it is asking for a name actually. we fixed that in core estimators.

**Comment by brightbytes-dude** [comment]
Yup, I'm providing it with a column name, and reshaping my label solved the crash I was getting.  

As described in my comments above, assuming I stay with the contrib Estimator, the problem is now that I need to figure out how to test a trained model without the weight feature being present, and also how to use a trained model for predictions without that weight feature being present.

Or, are you saying that the contrib DNNClassifier `weight_column_name` param just doesn't work at all?

Basically, I think you described the correct approach to solving the problem by stating that the weight should be processed in the input function.  That way, it never ends up in the features, and the model doesn't need to be told about it.

I just don't know how to do that ... yet. :-)  Any hints would be very much appreciated.

**Comment by brightbytes-dude** [comment]
I tried that, though with `tf.contrib.learn.DNNClassifier` rather than `tf.estimator.DNNClassifier`. (I don't think it matters which I use because both the `weight_column_name` of the former and the `weight_column` of the latter expect the column referenced to be a feature column.)

Specifically, if I:

1. Construct a `tf.contrib.learn.DNNClassifier`, and specify `weight_column_name` and include that column among the `feature_columns`.

2. Call its `fit` method with an `input_fn` that includes the weight column among the feature_columns.

3. Call its `evaluate` method with an `input_fn` that does **not** supply the weight column among the feature_columns.

... then `evaluate` crashes with an error that the weight column is missing.

If, instead, I try this approach:

1. Construct a `tf.contrib.learn.DNNClassifier`, and specify `weight_column_name` and include that column among the `feature_columns`.

2. Call its `fit` method with an `input_fn` that includes the weight column among the feature columns.

3. Construct a new `tf.contrib.learn.DNNClassifier` that uses the same `model_dir` as the classifier created in step 1, but that does not specify the `weight_column_name` param and does not include the weight column among the `feature_columns`.

4. Call the `evaluate` method of the new classifier with an `input_fn` that does **not** include the weight column among the feature columns.

... then it crashes, as I described in this [StackOverflow](https://stackoverflow.com/a/46249103/915989) issue.

It seems to me that the requirement that the class weights column be one of the model's feature columns renders this feature entirely useless for remedying class imbalance.

(It's undoubtedly useful for other use cases; I'm just not currently interested in any of them.)  :-)  :-)

**Comment by ispirmustafa** [comment]
you should not include `weight_column`(`_name`) as part of `feature_columns`. Providing that as part of `feature_columns` is saying that you're using it as a feature which is wrong.

**Comment by brightbytes-dude** [comment]
If I don't provide it as a feature column, I get an error that it needs to be provided as a feature column.

Also, here's the doc for the param of `tf.contrib.learn.DNNClassifier`, with bolding added:

```
weight_column_name: A string defining **feature column** name representing weights. It is used to down weight or boost examples during training. It will be multiplied by the loss of the example.
```

... and, here's the doc for the param of `tf.estimator.DNNClassifier`, with bolding added:

```
weight_column: A string or a _NumericColumn created by tf.feature_column.numeric_column defining **feature column** representing weights. It is used to down weight or boost examples during training. It will be multiplied by the loss of the example. If it is a string, it is used as a key to fetch weight tensor **from the features**. If it is a _NumericColumn, raw tensor is fetched by key weight_column.key, then weight_column.normalizer_fn is applied on it to get weight tensor.
```

Again, I tried not defining it as a feature column, hoping that documentation was in error.  But, it's not.

**Comment by ispirmustafa** [comment]
it would be great if you can create a small example where you see the problem. then we can discuss the code of that example.

**Comment by ednaruiz** [comment]
did someone find the solution to this? :-) 
@brightbytes-dude you maybe? I have exactly the problem you describe up here.

**Comment by brightbytes-dude** [comment]
@ednaruiz - I still owe the thread some failing `pytest`s to demonstrate the issues with various attempted workarounds. (I haven't found the time to yet, and sadly am now working on another project for a few weekds.)

Ultimately though, the only solutions I could come up with are:

1. Over-sample the under-represented class's examples.  

2. Use the low-level API to define the loss function by hand such that it uses the contents of the weight column when determining loss, and do not define the weight column as a feature column.  (I'm not sure how to do this yet, but there seems enough data on the web that I think I could figure it out.)

To reiterate the problem: if one attempts to solve the Class Imbalance problem by defining a weight column as a feature column when the class weights correspond to the class, then the network can trivially attain perfect accuracy simply by 100% correlating the weight to the outcome.

For example, say I'm building a dog classifier and I have 100 pictures, 95 of which are **not** dogs, and the remaining 5 are dogs.

I then add a weight column to my dataset, where the not-dog pictures have a class weight of 1/95 and the dog pictures have a weight of 1/5, and I specify the weight column when constructing my model.

By doing so, the weight column becomes a feature column.  Because it's a feature column, my input function needs to supply values for it, and my model will use it to "guess" the correct class.  

I then train my model, and find that no matter how bad the dog pictures are, every training run always achieves 100% accuracy.  

That is, because the weight column is a feature column, the model trivially learns that if the class weight is 1/5, it's a dog, and if the class weight is 1/95, it's not a dog. Viva la ML!! :-)

Furthermore, since my test runs and prediction runs require that the column be present (because it's a feature column of the model), I'm stuck specifying **something** when testing and predicting.  

For my actual model, hoping I could work around this, I tried setting the weight column value for testing and predicting to first the mean and then the midrange, but in both cases the trained models had worse accuracy than if I'd never used the column weights param.  So, I gave up.

My conclusion is that one can only use the `weight_column_name` parameter when the values in that column are **not** correlated with the outcome.  

However, to solve the Class Imbalance problem, the `weight_column_name` must be correlated with the outcome.  So, I can't use this param in my modeling.

**Comment by tilarids** [comment]
@brightbytes-dude [testLossWithWeights in dnn_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/dnn_test.py#L632) is testing the case you're describing. Check it out to see how you can use weight column in your input_fn without adding it to DNNClassifier feature_columns. You mentioned that you've tried exactly that but it seems there may be just a simple bug in your code. Hope this helps locating the issue.

**Comment by bizstream-markschmidt** [comment]
@brightbytes-dude I think you are experiencing the same exact issue I am https://stackoverflow.com/questions/48460913/how-do-i-properly-use-the-weight-column-when-working-with-tf-estimator-dnnclassi

Did you ever figure out a solution to it?

**Comment by brightbytes-dude** [comment]
@tilarids - Sorry for the (very!) long delay: I wasn't able to get back to the project until recently.

The test case you indicated allowed me to get this working on my project: thank you very much for referencing it!!

My mistake was to take the documentation literally when it says `weight_column_name: A string defining feature column name representing weights.`.  I assumed that meant that `weight_column_name` - `w` in the test - **must** be one of the model features.  Whereas, the test case clearly shows that it absolutely should not be.

And, using weights had exactly the effect I'd hoped: it improved training & test Recall at 50% threshold in my binary classification problem, though of course at the expense of reducing Accuracy.  But AUC remains the same whether or not weights are used, demonstrating to my satisfaction that the TF implementation of case weighting is correct.

Finally, I have one question about that test case: in that example, one of the the prediction cases for evaluation has a different weight than the rest.  What is the actual use case for having a batch of predictions where the individual cases have different weights associated with them?

I ask because in my domain, all cases to be labelled with a prediction matter equally.  So, I was planning on just passing in the exact same weight for every single one of them, for every batch.  Thus, the second question: does it matter what constant weight I choose?

Apologies in advance if I should be asking these questions elsewhere ...
